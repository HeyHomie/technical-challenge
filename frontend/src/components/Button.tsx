import clsx from 'clsx'

interface Props {
  children: React.ReactNode
  className?: string
  type?: 'primary' | 'green'
  onClick?: () => void
  onBlur?: () => void
  noBorderLeft?: boolean
  noBorderRight?: boolean
}

export function Button({
  children,
  className,
  type,
  noBorderLeft,
  noBorderRight,
  onClick,
  onBlur
}: Props) {
  const buttonClassNames =
    'w-full relative text-sm py-[5px] px-[16px] text-center font-medium whitespace-nowrap cursor-pointer align-middle select-none rounded-md border'

  if (type === 'green') {
    return (
      <button
        onClick={onClick}
        onBlur={onBlur}
        className={clsx(
          className,
          noBorderLeft && 'rounded-l-none',
          noBorderRight && 'rounded-r-none',
          'bg-[#2da44e] hover:bg-[#46954a] active:bg-[#347d39] border-[#cdd9e51a] hover:border-[#cdd9e51a] text-white',
          buttonClassNames
        )}>
        {children}
      </button>
    )
  }

  return (
    <button
      onClick={onClick}
      onBlur={onBlur}
      className={clsx(
        className,
        noBorderLeft && 'rounded-l-none',
        noBorderRight && 'rounded-r-none',
        'border-[#1b1f2426] hover:border-[#1b1f2426] dark:border-[#cdd9e51a] dark:hover:border-[#768390] text-[#24292f] dark:text-[#adbac7] bg-[#f6f8fa] hover:bg-[#f3f4f6] active:bg-[#ebecf0] dark:bg-[#373e47] dark:hover:bg-[#444c56] dark:active:bg-[#3d444d]',
        buttonClassNames
      )}>
      {children}
    </button>
  )
}
