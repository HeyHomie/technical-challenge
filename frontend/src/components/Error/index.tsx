import React from 'react'

export const Error = (): JSX.Element => {
  return <h1>Error</h1>
}
