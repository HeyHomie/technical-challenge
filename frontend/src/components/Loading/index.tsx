import React from 'react'

export const Loading = (): JSX.Element => {
  return <h1>Loading</h1>
}
