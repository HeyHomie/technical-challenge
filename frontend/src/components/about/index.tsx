import { Link } from 'react-router-dom';
import './../../index.css'

export function About() {
    return (
        <div className='App'>
            <div style={{ margin: 'auto', width: '75%'}}>
                <h1>About Page</h1>

                I upload in this single commits all changes; but, How would my commits be?
                <ol>1) Migration removing jsonb field for repositories in users table</ol>
                <ol>2) Tests for models, after that you can improve severals db's features, example an index for name github_id email (This will add speed and restrictions)</ol>
                <ol>3) Change to services for users an repos creation and listing, with theirs tests</ol>
                <ol>4) A background job for repos saving with sidekiq and ActiveJob (this is great for an user with 1000 repos)</ol>
                <ol>5) A search with elasticsearch and searchkick</ol>
                <ol>6) Migration for aditional data</ol>
                <ol>7) Front-end Job (My focus is on back-end), but I have experience and I added a litle bit</ol>

                <Link to="/user/yknx4" >Home</Link>
            </div>
        </div>
    )
}