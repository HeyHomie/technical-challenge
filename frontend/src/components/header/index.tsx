interface IHeader {
    avatarUrl: string
}

import menuIcon from '../../assets/menu.png';
import logoIcon from '../../assets/logo.png';
import notificationIcon from '../../assets/notification.png';
import addIcon from '../../assets/plus.png';
import { Link } from 'react-router-dom';
import { useState } from 'react';

export function Header({ avatarUrl }: IHeader) {
    const [username, setUsername] = useState('');

    const getRepos = () => {
        window.open(`/user/${username && username.length >= 1 ? username : 'yknx4'}`, '_self');
    }

    return (
        <div id="header">
            <img src={menuIcon} alt="menu" id="menu-icon" />
            <div id="header-left">
                <div id="logo">
                    <img src={logoIcon} alt="logo" />
                </div>
                <div id="header-search">
                    <input placeholder="Search or jump to..." type="text" onChange={(e) => { setUsername(e.target.value) }} />
                    <button>/</button>
                </div>
                <div id="header-links">
                    <div className="link" onClick={() => getRepos()}>Search</div>
                    <div className="link"><Link to="/repos/*" className="white-links">All Repos</Link></div>
                    <div className="link"><Link to="/about" className="white-links">About</Link></div>
                </div>
            </div>
            <div className="flex"></div>
            <div id="header-right">
                <div id="header-icons">
                    <button id="notification-btn">
                        <img  className="icon" src={notificationIcon} alt="notification icon" />
                    </button>
                    <button id="plus-btn">
                        <img  className="icon" src={addIcon} alt="add icon" />
                        <span>&#9662;</span>
                    </button>
                    <button id="avatar-btn">
                        <img className="avatar" src={avatarUrl} alt="add icon" />
                        <span>&#9662;</span>
                    </button>
                </div>
            </div>
        </div>
    );
};