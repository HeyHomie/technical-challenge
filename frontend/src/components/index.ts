export * from './Nav'
export * from './Container'
export * from './Layout'
export * from './SideBar'
