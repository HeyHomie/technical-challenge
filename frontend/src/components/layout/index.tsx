import Main from './main'
import Navbar from './navbar'
import Footer from './footer'
import Grid from './grid'

export { Main, Navbar, Footer, Grid }
