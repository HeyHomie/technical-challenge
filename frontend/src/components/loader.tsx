import './loader.css'
import { Header, Icon } from "semantic-ui-react";

const Loader = () => (
  <div className='loader'>
    <Icon loading name='asterisk' size='massive' />
    <Header as='h1'>Loading...</Header>
  </div>
)

export default Loader