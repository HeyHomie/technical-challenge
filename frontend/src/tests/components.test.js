import TestRenderer from 'react-test-renderer'
import { Repo, Loader, Profile, Footer, Navbar, Paginator, Search } from '../components'

describe('Components', () => {
  it('renders repo component correctly', () => {
    const repo = {
      id: 1,
      github_id: 1,
      url: 'github.com/testuser/testrepo',
      name: 'testrepo',
      user_id: 1,
      fork: false,
      description: 'This is a test repo',
      language: null,
      stars: 0,
      forks: 1,
      license: 'MIT',
      last_updated: '2021-10-04 02:07:07',
      archived: false,
      private: false,
      created_at: '2021-10-04 02:07:07',
      updated_at: '2021-10-04 02:07:07'
    }

    const tree = TestRenderer
      .create(<Repo repo={repo}/>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders loader component correctly', () => {
    const tree = TestRenderer
      .create(<Loader/>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders profile component correctly', () => {
    const user = {
      id: 1,
      login: 'TestUser',
      github_id: 1,
      url: 'github.com/testuser',
      name: 'User',
      email: 'test@user.com',
      avatar_url: 'github.com/testuser',
      bio: null,
      company: null,
      location: null,
      blog: null,
      twitter_username: null,
      followers: 0,
      following: 0,
      created_at: '2021-10-04 02:07:07',
      updated_at: '2021-10-04 02:07:07'
    }

    const tree = TestRenderer
      .create(<Profile user={user}/>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders footer component correctly', () => {
    const tree = TestRenderer
      .create(<Footer/>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders navbar component correctly', () => {
    const tree = TestRenderer
      .create(<Navbar count={1}/>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders paginator component correctly', () => {
    const pageInfo = {
      page: 1,
      perPage: 5,
      totalCount: 1,
      totalPages: 1
    }

    const tree = TestRenderer
      .create(
        <Paginator 
        pageInfo={pageInfo} 
        handleChange={() => {alert('Paginator test!')}}/>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders search component correctly', () => {
    const tree = TestRenderer
      .create(<Search page={1} onSearch={() => {alert('Search test!')}}/>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
