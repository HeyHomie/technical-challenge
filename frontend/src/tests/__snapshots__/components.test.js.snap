// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components renders footer component correctly 1`] = `
Array [
  <div
    className="ui divider"
  />,
  <div
    className="ui gh-footer menu"
  >
    <div
      className="item"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="grey copyright outline icon"
        onClick={[Function]}
      />
      <span>
        2021 GitHub, Inc.
      </span>
    </div>
    <a
      className="item"
      href="https://docs.github.com/en/github/site-policy/github-terms-of-service"
      onClick={[Function]}
      target="_blank"
    >
      Terms
    </a>
    <a
      className="item"
      href="https://docs.github.com/en/github/site-policy/github-privacy-statement"
      onClick={[Function]}
      target="_blank"
    >
      Privacy
    </a>
    <a
      className="item"
      href="https://github.com/security"
      onClick={[Function]}
      target="_blank"
    >
      Secutiry
    </a>
    <a
      className="item"
      href="https://www.githubstatus.com/"
      onClick={[Function]}
      target="_blank"
    >
      Status
    </a>
    <a
      className="item"
      href="https://docs.github.com/"
      onClick={[Function]}
      target="_blank"
    >
      Docs
    </a>
    <a
      className="item grey"
      href="https://github.com/"
      onClick={[Function]}
      target="_blank"
    >
      <i
        aria-hidden="true"
        className="grey github big icon"
        onClick={[Function]}
      />
    </a>
    <a
      className="item"
      href="https://support.github.com/?tags=dotcom-footer"
      onClick={[Function]}
      target="_blank"
    >
      Contact Git Hub
    </a>
    <a
      className="item"
      href="https://github.com/pricing"
      onClick={[Function]}
      target="_blank"
    >
      Pricing
    </a>
    <a
      className="item"
      href="https://docs.github.com/"
      onClick={[Function]}
      target="_blank"
    >
      API
    </a>
    <a
      className="item"
      href="https://services.github.com/"
      onClick={[Function]}
      target="_blank"
    >
      Training
    </a>
    <a
      className="item"
      href="https://github.blog/"
      onClick={[Function]}
      target="_blank"
    >
      Blog
    </a>
    <a
      className="item"
      href="https://github.com/about"
      onClick={[Function]}
      target="_blank"
    >
      About
    </a>
  </div>,
]
`;

exports[`Components renders loader component correctly 1`] = `
<div
  className="loader"
>
  <i
    aria-hidden="true"
    className="asterisk massive loading icon"
    onClick={[Function]}
  />
  <h1
    className="ui header"
  >
    Loading...
  </h1>
</div>
`;

exports[`Components renders navbar component correctly 1`] = `
<div
  className="ui large pointing secondary top fixed gh-menu menu"
>
  <div
    className="link item gh-item"
    onClick={[Function]}
  >
    <i
      aria-hidden="true"
      className="grey newspaper outline icon"
      onClick={[Function]}
    />
    <span>
      Overview
    </span>
  </div>
  <div
    className="active link item gh-item-active"
    onClick={[Function]}
  >
    <i
      aria-hidden="true"
      className="grey book icon"
      onClick={[Function]}
    />
    <span>
      Repositories
    </span>
    <div
      className="ui grey label"
      onClick={[Function]}
    >
      1
    </div>
  </div>
  <div
    className="link item gh-item"
    onClick={[Function]}
  >
    <i
      aria-hidden="true"
      className="grey trello icon"
      onClick={[Function]}
    />
    <span>
      Projects
    </span>
  </div>
  <div
    className="link item gh-item"
    onClick={[Function]}
  >
    <i
      aria-hidden="true"
      className="grey box icon"
      onClick={[Function]}
    />
    <span>
      Packages
    </span>
  </div>
</div>
`;

exports[`Components renders paginator component correctly 1`] = `
<div
  aria-label="Pagination Navigation"
  className="ui pagination pointing secondary gh-paginator menu"
  role="navigation"
>
  <a
    aria-current={false}
    aria-disabled={false}
    aria-label="Previous item"
    className="item"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
    type="prevItem"
    value={1}
  >
    ⟨
  </a>
  <a
    aria-current={true}
    aria-disabled={false}
    className="active item"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
    type="pageItem"
    value={1}
  >
    1
  </a>
  <a
    aria-current={false}
    aria-disabled={false}
    aria-label="Next item"
    className="item"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
    type="nextItem"
    value={1}
  >
    ⟩
  </a>
</div>
`;

exports[`Components renders profile component correctly 1`] = `
Array [
  <img
    className="ui medium circular image"
    src="github.com/testuser"
  />,
  <h2
    className="ui header"
  >
    User
  </h2>,
  <h3
    className="ui header"
  >
    TestUser
  </h3>,
  <button
    className="ui fluid button"
    onClick={[Function]}
  >
    Follow
  </button>,
  <p
    className="bio"
  />,
  <span
    className="stats"
  >
    <i
      aria-hidden="true"
      className="users icon"
      onClick={[Function]}
    />
    <span
      className="number"
    >
       0 
    </span>
     followers
  </span>,
  <span
    className="stats"
  >
     
    <span
      className="number"
    >
      0
    </span>
     following 
  </span>,
  <span
    className="stats"
  >
    <i
      aria-hidden="true"
      className="star outline icon"
      onClick={[Function]}
    />
    <span
      className="number"
    >
      38
    </span>
  </span>,
  <div
    className="social-container"
  >
    <p
      className="social"
    >
      <i
        aria-hidden="true"
        className="mail outline icon"
        onClick={[Function]}
      />
      <a
        href="mailto:test@user.com"
      >
        test@user.com
      </a>
    </p>
  </div>,
]
`;

exports[`Components renders repo component correctly 1`] = `
<div>
  <div
    className="ui grid"
  >
    <div
      className="row"
    >
      <div
        className="twelve wide column"
      >
        <div
          className="title-container"
        >
          <a
            className="ui header repo"
            href="github.com/testuser/testrepo"
            target="_blank"
          >
            testrepo
          </a>
          <div
            className="ui circular label"
            onClick={[Function]}
          >
            Public
          </div>
        </div>
        <div
          className="details-container"
        >
          <div
            className="details-item"
          >
            <i
              aria-hidden="true"
              className="fork icon"
              onClick={[Function]}
            />
            <span>
              1
            </span>
          </div>
          <div
            className="details-item"
          >
            <i
              aria-hidden="true"
              className="balance icon"
              onClick={[Function]}
            />
            <span>
              MIT
            </span>
          </div>
          <div
            className="details-item"
          >
            <span>
              Updated on 
              Oct 4
            </span>
          </div>
        </div>
      </div>
      <div
        className="right floated four wide column"
      >
        <button
          className="ui small compact right floated button"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="star outline icon"
            onClick={[Function]}
          />
          <span>
            Star
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    className="ui divider"
  />
</div>
`;

exports[`Components renders search component correctly 1`] = `
<div
  className="gh-search"
>
  <div
    className="ui stackable grid"
  >
    <div
      className="row"
    >
      <div
        className="eight wide column"
      >
        <div
          className="ui fluid input"
        >
          <input
            onChange={[Function]}
            placeholder="Find a repository..."
            type="text"
          />
        </div>
      </div>
      <div
        className="eight wide column"
      >
        <div
          className="gh-filters"
        >
          <div
            aria-expanded={false}
            className="ui selection dropdown"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="listbox"
            tabIndex={0}
          >
            <div
              aria-atomic={true}
              aria-live="polite"
              className="divider default text"
              role="alert"
            >
              Type
            </div>
            <i
              aria-hidden="true"
              className="dropdown icon"
              onClick={[Function]}
            />
            <div
              className="menu transition"
            >
              <div
                aria-checked={true}
                aria-selected={true}
                className="active selected item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  All
                </span>
              </div>
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  Forks
                </span>
              </div>
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  Archived
                </span>
              </div>
            </div>
          </div>
          <div
            aria-expanded={false}
            className="ui selection dropdown"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="listbox"
            tabIndex={0}
          >
            <div
              aria-atomic={true}
              aria-live="polite"
              className="divider default text"
              role="alert"
            >
              Language
            </div>
            <i
              aria-hidden="true"
              className="dropdown icon"
              onClick={[Function]}
            />
            <div
              className="menu transition"
            >
              <div
                aria-checked={true}
                aria-selected={true}
                className="active selected item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  All
                </span>
              </div>
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  HTML
                </span>
              </div>
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  TypeScript
                </span>
              </div>
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  JavaScript
                </span>
              </div>
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  Ruby
                </span>
              </div>
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  CSS
                </span>
              </div>
            </div>
          </div>
          <div
            aria-expanded={false}
            className="ui selection dropdown"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="listbox"
            tabIndex={0}
          >
            <div
              aria-atomic={true}
              aria-live="polite"
              className="divider default text"
              role="alert"
            >
              Sort
            </div>
            <i
              aria-hidden="true"
              className="dropdown icon"
              onClick={[Function]}
            />
            <div
              className="menu transition"
            >
              <div
                aria-checked={true}
                aria-selected={true}
                className="active selected item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  Last updated
                </span>
              </div>
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  Name
                </span>
              </div>
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  Stars
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ui divider"
  />
</div>
`;
